### DISCUSSION TITLE
Very Noisy Gradients with Respect to Latent Variables?

### DISCUSSION THREAD

**Discussion by coult099** [mitsuba-renderer/mitsuba3#1215] [Q&A] [answered]
Hello,

I am having an issue with forward inverse rendering: when differentiating with respect to latent variables in the scene, the gradient image appears much noisier than otherwise. I have included an example with images below, based on the Cornell box scene. Here, the small ball is red and all walls are white. 

Is there a reason why the latent variable gradients (size, position) appear so much noisier than the non-latent variable gradient (color)? Is the gradient with respect to these parameters just so small that noise dominates? Or, is there an issue with my implementation? 

A confounding factor: I have repeated the experiment with much higher SPP and while the gradient does get cleaner with more samples (although not as clean as the non-latent gradient), some artifacts/bright pixels start to appear in the gradient image, as seen below. These bright pixels/artifacts also tend to appear in the same locations on successive runs, sometimes even with different random seeds. Thus, while I would otherwise try to average out the gradient over successive runs to decrease the noise, I can't do that here since the artifacts sometimes constructively interfere. Any help is appreciated! 

System:
- macOS Sonoma 14.5 (Apple Silicon)
- Python 3.10.8 (running in a virtual environment)
- Mitsuba 3.5.2
- Drjit 0.4.6
- PRB integrator with `max_depth=-1` (infinite bounces).
- Sampler: `independent`
 
```py
import drjit as dr
import mitsuba as mi

mi.set_variant('llvm_ad_rgb')
scene = mi.load_file('./scenes/cbox/cbox.xml')
params = mi.traverse(scene)

# From Cornell box .xml
spp_val = 1024
ball_size = mi.Float(0.25)
ball_x = mi.Float(0.5)
ball_y = mi.Float(-0.75)
ball_z = mi.Float(-0.2)
pos_add = mi.Float(0)
gradient_param = "size" # "size", "color", or "position"

if gradient_param == "color":
    key = 'red.reflectance.value'
    dr.enable_grad(params[key])
    params.update();
elif gradient_param == "size":
    key = "redsphere.to_world"
    dr.enable_grad(ball_size)
    params[key] = mi.Transform4f(mi.Matrix4f([[ball_size, 0, 0, ball_x + pos_add], [0, ball_size, 0, ball_y + pos_add], [0, 0, ball_size, ball_z + pos_add], [0, 0, 0, 1]]))
    params.update();
elif gradient_param == "position":
    key = 'redsphere.to_world'
    dr.enable_grad(pos_add)
    params[key] = mi.Transform4f(mi.Matrix4f([[ball_size, 0, 0, ball_x + pos_add], [0, ball_size, 0, ball_y + pos_add], [0, 0, ball_size, ball_z + pos_add], [0, 0, 0, 1]]))
    params.update();

if gradient_param == "size":
    dr.forward(ball_size, dr.ADFlag.ClearEdges)
elif gradient_param == "position":
    dr.forward(pos_add, dr.ADFlag.ClearEdges)

# Render image
image = mi.render(scene, params, spp=spp_val)

if gradient_param == "color":
    dr.forward(params[key])
    grad_image = dr.grad(image)
else:
    dr.forward_to(image)
    grad_image = 2 * dr.grad(image)

mi.util.write_bitmap('cbox_original.png', image)
mi.util.write_bitmap('cbox_'+gradient_param+'_grad.png', grad_image)
```

SPP = 1024:
|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/9feb5f4d-9288-4d4d-a248-549137d0462e)|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/575e0e39-262b-40d2-952c-2d4d9ae4b14d)|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/e2b325cd-6082-4d77-b712-d2a13adee2b8)|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/62057b0b-0049-4860-8d29-0fde22917174)|
|:-:|:-:|:-:|:-:|
|Original Image|Red Ball Color Gradient|Red Ball Size Gradient|Red Ball Position Gradient|

SPP = 16384:
|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/e57b4d22-cc85-47e4-bfc2-d0c30f50f8d7)|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/4f622bab-6f80-4ec8-b5d1-17e132f91b87)|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/e7f0da25-c693-4cd1-8daf-1d5fee3304fe)|![](https://github.com/mitsuba-renderer/mitsuba3/assets/115170151/3eaaeffd-f09c-4516-b79a-f32dc7d65331)|
|:-:|:-:|:-:|:-:|
|Original Image|Color Gradient|Size Gradient|Position Gradient|

**Comment by njroussel** [comment, marked as answer]
Hi @coult099 

Your code looks fine.

I do indeed think that the noise-level difference is also related to the magnitude of the actual derivative. Indirect diffuse light will barely see any change if you scale/move the red ball.

However, I do thing that your gradients for scale and position are wrong. Are you using the `prb_projective` integator? Unless you're using a `*_projective` integrator, you won't be accounting for the gradients that arise from changing visibility. (There's a small introduction on this topic in [this tutorial](https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/object_pose_estimation.html).) You can double-check your forward gradients by comparing the to a "brute-force" finite-differences computation, similarly to what is done in [this tutorial](https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/projective_sampling_integrators.html).

**Reply by coult099** [reply]
Hello @njroussel,

Thank you for the reply. I switched to using a projective integrator instead of the basic `prb` integrator, and that helped a bit: the gradients look close to a finite-difference gradient (for size & shape). However, running with `prb_projective` takes much longer than with `prb`, several orders of magnitude longer using `llvm_ad_rgb`. Is there any solution to this?

Additionally, when I try to run the `prb_projective` integrator with `cuda_ad_rgb`, it fails when forward-propagating gradients to the image, this step:
```py
dr.forward_to(image)
```
gives:
```
[list of parameters]
...
COMPILE ERROR: Invalid PTX input: ptx2llvm-module-002: error: Failed to parse input PTX string ptx2llvm-module-002, line 14278; error.     : Feature 'activemask' requires PTX ISA .version 6.2 or later
Cannot parse input PTX string


Aborted (core dumped)
```

Do you have any idea why this might be happening? Thank you!

**Reply by coult099** [reply]
@njroussel,

The issue above appears to be a sample issue. I repeated the same experiment using the `independent` sampler instead of `ldsampler` and got the familiar printout
```
jit_flush_malloc_cache(): Dr.Jit exhausted the available memory and had to flush its allocation cache to free up additional memory. This is an expensive operation and will have a negative effect on performance. You may want to change your computation so that it uses less memory. This warning will only be displayed once.
``` 

However, I am not requiring a large image: 1024 SPP at 256 x 256 resolution in RGB is only ~2^20 samples, not near the 2^32 limit for Dr.Jit. So what is happening here with the gradient propagation? And, why is it only causing issues with `cuda_ad_rgb`, not with `llvm_ad_rgb`?

To add to this: I do not think it is a hardware issue: I am running on a Quadro RTX 8000 with 48GB of memory.

**Reply by njroussel** [reply]
`prb_projective` needs to compute a complex data structure in order to compute gradients correctly, whereas `prb` does not need any of this but, as mentioned before, it will miss some gradients.

The error message above indicates that you've ran out of memory. The aforementioned data structure lives in memory, that is what is most likely eating up all your memory. 
There are some hyperparameters to control the size and accuracy of the data structure: https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/common.py#L826-L895
You can play with these to try to get everything to fit. However, we picked the defaults to be fairly reasonable for GPUs with much less VRAM than what you have. So I'm inclined to believe you have something else that is using a good chunk of your memory.

**Reply by coult099** [reply]
Hello @njroussel,

Thank you again for the reply. I was able to make sure nothing else was running on the RTX cards, and got the gradient renders working. Thank you for the help.

One final thing: the intensities in the finite-difference vs automatic differentiation gradients were pretty different from one another. Also, there are some artifacts which show up in the finite-difference gradient for size which are not present in the AD gradient image (below).

Here are the final images I was able to generate (SPP = 4096):
| **Auto-diff Gradient** |![](https://github.com/user-attachments/assets/242276df-fe4a-4687-a035-a22f54cd9c68)|![](https://github.com/user-attachments/assets/2d2cd57f-fff7-4b7a-a6c0-1b18bfd03c63)|
|:-:|:-:|:-:|
| **Finite-Difference Gradient** |![](https://github.com/user-attachments/assets/5bce22d5-db8f-4742-88e9-3843ec0e26f7)|![](https://github.com/user-attachments/assets/a65b16d6-a9fa-433d-82bf-8a7f22bd8fd5)|
| |Red Ball Position | Red Ball Size ( `log(abs(image))` ) |

**Reply by njroussel** [reply]
What material is used for the shiny ball? It cannot be a perfect mirror, the method cannot support BSDF which Dirac delta lobes.

**Reply by coult099** [reply]
I did not change the shiny ball material from the default: `<bsdf type="conductor" id="mirror"/>` . I will re-try the experiment with a more diffuse BSDF.

**Reply by njroussel** [reply]
Aha, that is the issue I believe.

**Reply by coult099** [reply]
Hello @njroussel ,

I conducted the test using the `roughplastic` BSDF in green for the larger sphere, and got more similar results, as well as a faster runtime. Original image
![cbox_original_4096](https://github.com/user-attachments/assets/e6c69f20-c20d-4d3e-b84a-7cd330c00c9f)


Gradients (SPP = 4096):
| **Auto-diff Gradient** |![](https://github.com/user-attachments/assets/e55246e5-af76-4168-8e27-84d33ad13d5c)|![](https://github.com/user-attachments/assets/3483e58f-f675-4a66-b481-2fde930e198c)|
|:-:|:-:|:-:|
| **Finite-Difference Gradient** |![](https://github.com/user-attachments/assets/4e914802-aa85-4f7a-b14a-14bbfc0eb8ab)|![](https://github.com/user-attachments/assets/92b67dbd-37a8-45b4-994f-02467ea42efc)|
| |Red Ball Position | Red Ball Size ( `log(abs(image))` ) |

The automatic differentiation gradient for both position and size contains more noise (and more spurious pixel groups, as in my very first post) than the finite-difference gradient. Do you know why this would be? I generated the auto-diff gradients with 4096 SPP, and the FD gradients with 32 rounds of 4096 SPP each.

**Reply by coult099** [reply]
I re-ran these experiments with higher SPP for the AD gradients, and got a similar output. In the higher SPP renders, there are spurious pixel groups/artifacts in the AD images, which I would expect not to be the case (indeed, this is the initial issue I had). Because of these artifacts, the FD gradient looks less noisy than the AD gradient, something I would not at all expect to be the case. Do you know what might be causing this behavior? I am using a delta size of `1e-3` for the FD gradients.

Gradients:
|**AD: 4096 SPP**|![](https://github.com/user-attachments/assets/12f69fae-723f-4873-94b4-4ef2747e462a)|![](https://github.com/user-attachments/assets/239a672f-b63a-4aa7-a332-769ada37e71f)|![](https://github.com/user-attachments/assets/d2d9d44b-0060-43dc-aabc-9fc48c613ae7)|
|:-:|:-:|:-:|:-:|
| **AD: 16384 SPP** |![](https://github.com/user-attachments/assets/3641e59b-df62-40e5-a423-ed51ed467ce9)|![](https://github.com/user-attachments/assets/5294ad53-82b6-4f78-99a0-c228b337ed46)|![](https://github.com/user-attachments/assets/f445a298-c2e4-42ff-a63f-9f90fed1482f)|
| **FD: 32 x 4096 SPP** |![](https://github.com/user-attachments/assets/23c96cbb-bcf6-46ca-ad3d-333cf47fc1bb)|![](https://github.com/user-attachments/assets/409bed11-10c4-4162-844a-6acb5a208266)|![](https://github.com/user-attachments/assets/317aa4b1-86e3-4ea5-8910-7ac93716af3a)|
|     |Color Gradient|Size Gradient|Position Gradient|

**Reply by coult099** [reply]
I've repeated the experiment with different scenes, and always seem to get the same result: color gradients are very clean, but position and size gradients are always noisy with spurious groups of pixels/artifacts as seen above. It is very surprising and counter-intuitive that using finite differences can give a cleaner gradient image than automatic differentiation! I hope I am doing something incorrectly.

**Reply by njroussel** [reply]
Hi @coult099 

> color gradients are very clean, but position and size gradients are always noisy 

This is expected. Measuring gradients induced by visibility discontinuities is a difficult problem, these only appear in the latter case when scene geometry is changing.

>  It is very surprising and counter-intuitive that using finite differences can give a cleaner gradient image than automatic differentiation!

Not really. We're not just running AD, these are full on reverse Monte Carlo simulations, so you can expect the same noise characteristics as traditional forward rendering. In addition, with visibility changes, the space of paths that produce forward gradients is extremely sparse.

I obviously can't guarantee that your code is bug-free, but from the results you've shown I'd say that the results are as expected. Keep in mind that in order to run an optimization you don't need perfectly clean gradients, low gradient spp renderings work fine in most cases.
